File: QolmsYappli/QolmsYappli/Models/PartialViewModel/HealthAgeUrineReportPartialViewModel.vb
Imports System.Collections.ObjectModel

<Serializable()>
Public NotInheritable Class HealthAgeUrineReportPartialViewModel
    Inherits QyHealthAgeReportPartialViewModelBase

#Region "Constant"

    Private Shared ReadOnly valueSetN As New Dictionary(Of Decimal, String)() From {
        {1D, "－"},
        {2D, "±"},
        {3D, "＋"},
        {4D, "＋＋"},
        {5D, "＋＋＋"}
    }

#End Region

#Region "Variable"

    Private _tableData As New List(Of Tuple(Of String, String))()

#End Region

#Region "Public Property"

    Public ReadOnly Property TableData As ReadOnlyCollection(Of Tuple(Of String, String))

        Get
            Return Me._tableData.ToList().AsReadOnly()
        End Get

    End Property

#End Region

#Region "Constructor"

    Public Sub New(model As HealthAgeViewModel, reportItem As HealthAgeReportItem)

        MyBase.New(model, reportItem)

        Me.InitializeBy(reportItem)

    End Sub

#End Region

#Region "Private Method"

    Private Function ToStringValue(value As Decimal) As String

        Dim result As String = String.Empty

        If HealthAgeUrineReportPartialViewModel.valueSetN.ContainsKey(value) Then result = HealthAgeUrineReportPartialViewModel.valueSetN(value)

        Return result

    End Function

#End Region

#Region "Protected Method"

    Protected Overrides Sub InitializeBy(reportItem As HealthAgeReportItem)

        Me._tableData = New List(Of Tuple(Of String, String))()

        If reportItem IsNot Nothing _
            AndAlso reportItem.HealthAgeReportType = QyHealthAgeReportTypeEnum.Urine _
            AndAlso reportItem.HealthAgeValueN IsNot Nothing _
            AndAlso reportItem.HealthAgeValueN.Any() Then

            ' 尿糖
            Dim ch037N As List(Of HealthAgeValueItem) = reportItem.HealthAgeValueN.Where(Function(i) i.HealthAgeValueType = QyHealthAgeValueTypeEnum.Ch037).OrderByDescending(Function(i) i.RecordDate).Take(3).ToList()

            Select Case ch037N.Count
                Case 2
                    ch037N.Add(New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue})

                Case 1
                    ch037N.AddRange(
                        {
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue},
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue}
                        }
                    )

                Case 0
                    ch037N.AddRange(
                        {
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue},
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue},
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue}
                        }
                    )

            End Select

            ch037N.Reverse()

            ' 尿蛋白
            Dim ch039N As List(Of HealthAgeValueItem) = reportItem.HealthAgeValueN.Where(Function(i) i.HealthAgeValueType = QyHealthAgeValueTypeEnum.Ch039).OrderByDescending(Function(i) i.RecordDate).Take(3).ToList()

            Select Case ch039N.Count
                Case 2
                    ch039N.Add(New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue})

                Case 1
                    ch039N.AddRange(
                        {
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue},
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue}
                        }
                    )

                Case 0
                    ch039N.AddRange(
                        {
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue},
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue},
                            New HealthAgeValueItem() With {.RecordDate = Date.MinValue, .Value = Decimal.MinValue}
                        }
                    )

            End Select

            ch039N.Reverse()

            Me._tableData.Add(New Tuple(Of String, String)(Me.ToStringValue(ch037N(0).Value), Me.ToStringValue(ch039N(0).Value)))
            Me._tableData.Add(New Tuple(Of String, String)(Me.ToStringValue(ch037N(1).Value), Me.ToStringValue(ch039N(1).Value)))
            Me._tableData.Add(New Tuple(Of String, String)(Me.ToStringValue(ch037N(2).Value), Me.ToStringValue(ch039N(2).Value)))
        Else
            Me._tableData.Add(New Tuple(Of String, String)(String.Empty, String.Empty))
            Me._tableData.Add(New Tuple(Of String, String)(String.Empty, String.Empty))
            Me._tableData.Add(New Tuple(Of String, String)(String.Empty, String.Empty))
        End If

    End Sub

#End Region

End Class

File: QolmsJotoWebView/Models/PartialViewModel/HealthAgeUrineReportPartialViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace MGF.QOLMS.QolmsJotoWebView
{
    [Serializable]
    public sealed class HealthAgeUrineReportPartialViewModel : QjHealthAgeReportPartialViewModelBase
    {
        #region Constant

        private static readonly Dictionary<decimal, string> valueSetN = new Dictionary<decimal, string>
        {
            { 1m, "－" },
            { 2m, "±" },
            { 3m, "＋" },
            { 4m, "＋＋" },
            { 5m, "＋＋＋" }
        };

        #endregion

        #region Variable

        private List<Tuple<string, string>> _tableData = new List<Tuple<string, string>>();

        #endregion

        #region Public Property

        public ReadOnlyCollection<Tuple<string, string>> TableData
        {
            get { return this._tableData.ToList().AsReadOnly(); }
        }

        #endregion

        #region Constructor

        public HealthAgeUrineReportPartialViewModel(HealthAgeViewModel model, HealthAgeReportItem reportItem)
            : base(model, reportItem)
        {
            this.InitializeBy(reportItem);
        }

        #endregion

        #region Private Method

        private string ToStringValue(decimal value)
        {
            string result = string.Empty;

            if (HealthAgeUrineReportPartialViewModel.valueSetN.ContainsKey(value)) result = HealthAgeUrineReportPartialViewModel.valueSetN[value];

            return result;
        }

        #endregion

        #region Protected Method

        protected override void InitializeBy(HealthAgeReportItem reportItem)
        {
            this._tableData = new List<Tuple<string, string>>();

            if (reportItem != null
                && reportItem.HealthAgeReportType == QjHealthAgeReportTypeEnum.Urine
                && reportItem.HealthAgeValueN != null
                && reportItem.HealthAgeValueN.Any())
            {
                // 尿糖
                List<HealthAgeValueItem> ch037N = reportItem.HealthAgeValueN
                    .Where(i => i.HealthAgeValueType == QjHealthAgeValueTypeEnum.Ch037)
                    .OrderByDescending(i => i.RecordDate)
                    .Take(3)
                    .ToList();

                switch (ch037N.Count)
                {
                    case 2:
                        ch037N.Add(new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue });
                        break;

                    case 1:
                        ch037N.AddRange(new[]
                        {
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue },
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue }
                        });
                        break;

                    case 0:
                        ch037N.AddRange(new[]
                        {
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue },
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue },
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue }
                        });
                        break;
                }

                ch037N.Reverse();

                // 尿蛋白
                List<HealthAgeValueItem> ch039N = reportItem.HealthAgeValueN
                    .Where(i => i.HealthAgeValueType == QjHealthAgeValueTypeEnum.Ch039)
                    .OrderByDescending(i => i.RecordDate)
                    .Take(3)
                    .ToList();

                switch (ch039N.Count)
                {
                    case 2:
                        ch039N.Add(new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue });
                        break;

                    case 1:
                        ch039N.AddRange(new[]
                        {
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue },
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue }
                        });
                        break;

                    case 0:
                        ch039N.AddRange(new[]
                        {
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue },
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue },
                            new HealthAgeValueItem { RecordDate = DateTime.MinValue, Value = decimal.MinValue }
                        });
                        break;
                }

                ch039N.Reverse();

                this._tableData.Add(new Tuple<string, string>(this.ToStringValue(ch037N[0].Value), this.ToStringValue(ch039N[0].Value)));
                this._tableData.Add(new Tuple<string, string>(this.ToStringValue(ch037N[1].Value), this.ToStringValue(ch039N[1].Value)));
                this._tableData.Add(new Tuple<string, string>(this.ToStringValue(ch037N[2].Value), this.ToStringValue(ch039N[2].Value)));
            }
            else
            {
                this._tableData.Add(new Tuple<string, string>(string.Empty, string.Empty));
                this._tableData.Add(new Tuple<string, string>(string.Empty, string.Empty));
                this._tableData.Add(new Tuple<string, string>(string.Empty, string.Empty));
            }
        }

        #endregion
    }
}
